### PR TITLE
Add yolov3detectionoutput test and AVX optimization

### PR DESCRIPTION
**Pull request by zhiliu6**

### CODE REVIEW

**Comment by tencent-adm** [comment]
[CLA assistant check](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=1994) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=1994) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: zhiliu6<br/>:x: restyled-commits<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/ncnn?pullRequest=1994) it.</sub>

File: tests/test_yolov3detectionoutput.cpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    pd.set(6, anchors_scale);
+
+    std::vector<ncnn::Mat> weights(0);
+
+    int ret = test_layer<ncnn::Yolov3DetectionOutput>("Yolov3DetectionOutput", pd, weights, a);
+    if (ret != 0)
+    {
+        fprintf(stderr, "test_yolov3detectionoutput failed a.dims=%d a=(%d %d %d) ", a[0].dims, a[0].w, a[0].h, a[0].c);
+        fprintf(stderr, " num_class=%d num_box=%d", num_class, num_box);
+        fprintf(stderr, " confidence_threshold=%f nms_threshold=%f", confidence_threshold, nms_threshold);
```
Comment (nihui) [inline tests/test_yolov3detectionoutput.cpp:38]: missing "\n" ?

File: tests/CMakeLists.txt [hunk 2]
```diff
@@ -72,3 +72,4 @@ ncnn_add_layer_test(UnaryOp)
 ncnn_add_layer_test(Mish)
 ncnn_add_layer_test(Swish)
 ncnn_add_layer_test(LSTM)
+ncnn_add_layer_test(yolov3detectionoutput)
```
Comment (nihui) [inline tests/CMakeLists.txt:75]: use the extract camelcase style class name here

File: src/layer/yolov3detectionoutput.h [hunk 3]
```diff
@@ -39,6 +39,19 @@ class Yolov3DetectionOutput : public Layer
     Mat anchors_scale;
     int mask_group_num;
     ncnn::Layer* softmax;
+
+public:
```
Comment (nihui) [inline src/layer/yolov3detectionoutput.h:43]: ```protected``` for utility functions

File: src/layer/yolov3detectionoutput.h [hunk 4]
```diff
@@ -39,6 +39,19 @@ class Yolov3DetectionOutput : public Layer
...
+    struct BBoxRect
+    {
+        float xmin;
+        float ymin;
+        float xmax;
+        float ymax;
+        int label;
+    };
+    void qsort_descent_inplace(std::vector<BBoxRect>& datas, std::vector<float>& scores, int left, int right) const;
+    void qsort_descent_inplace(std::vector<BBoxRect>& datas, std::vector<float>& scores) const;
```
Comment (nihui) [inline src/layer/yolov3detectionoutput.h:53]: The reason why the score was placed in a separate vector was to implement the bboxdata-independent template function. Here you can put the score member into BBoxRect struct to simplify the implementation of qsort.
It should be more memory efficient and fast.

File: src/layer/yolov3detectionoutput.h (hunk 3 above)
Comment (zhiliu6) [reply src/layer/yolov3detectionoutput.h:43]: These function will be used by child classes.

File: tests/CMakeLists.txt (hunk 2 above)
Comment (zhiliu6) [reply tests/CMakeLists.txt]: solved

**Comment by nihui** [comment]
Thanks for your contribution !

**Pull request merged**